doc/mlsys_paper/paper.tex

```diff
@@ -270,18 +270,22 @@ \section{Types of objective functions}
 %\end{equation}
 %\vspace*{-0.4em}
 %\noindent
-for any function $f(x)$ that takes some vector or matrix input $x$.  Ideally,
-the user should only need to specify what $f(x)$ is (or provide an
-implementation of a function that can compute $f(x)$).  But it is impossible to
-implement something both so generic and fast---for instance, gradient-based
-approaches converge far more quickly than gradient-free approaches (in general),
-so we must design an abstraction that is able to simultaneously generalize to
+for any function $f(x)$ that takes a vector or matrix input $x$.  Ideally,
+the user should only need to specify what $f(x)$ is, or provide an
+implementation of a function that can compute $f(x)$.  However, it is very difficult to
+implement something that is both generic and fast---for instance, gradient-based
+approaches typically converge far quicker than gradient-free approaches.
+As such, we must design an abstraction that is able to simultaneously generalize to
 many problem types, as well as take advantage of accelerations and
-optimizations.
+code optimizations.
+%% CS: in the above sentence, it's not clear what is meant by "we must design ..."
+%% CS: should it be "we provide ..." or "ensmallen provides ..." ?
+%% CS: The overall intent of the sentence is not clear -- to my eyes
+%% CS: it doesn't seem to gel well with its surrounding context.
```

Comment:
Right, I was thinking this also.  The idea was to present the ideal---that a user just writes $f(x)$---but then point out this is not realistic, and as a result we have this abstraction where users often need to write $f(x)$ and $f'(x)$.

Why don't we just ditch that and try something like this... instead of "Ideally... ...code optimizations" we could do this:

```
...for any function $f(x)$ that takes a vector or matrix input $x$.
In most cases, $f(x)$ will have special structure; one example might be that $f(x)$ is differentiable.  Therefore, the abstraction we have designed for {\tt ensmallen} takes advantage of this structure.  A user can also, for instance, provide an implementation of $f'(x)$, which in turn allows first-order gradient-based optimizers to be used.  This can allow significant speedups.
```

(Perhaps the last sentence or two could be improved.)

Reply:
Looks good.  I recommend changing
"... takes advantage of this structure. A user can also, for instance, provide an
implementation of $f'(x)$"
to:
"... can optionally take advantage of this structure. For example, in addition to $f(x)$, the user can provide an implementation of $f'(x)$"

The "can optionally" is important, as we want to be absolutely clear that the user is not forced to define $f'(x)$ in all cases.

Another recommended change to emphasise positive aspects.  Change:
"This can allow significant speedups."
to"
"This can often lead to significant speedups."